src/audio/module_adapter/module/intel_modules.c

```diff
@@ -0,0 +1,334 @@
+
+	return free_internal(mod_adp);
+}
+
+
+static int adp_wrapper_process(struct processing_module *mod,
+		struct input_stream_buffer *input_buffers,
+		int num_input_buffers, struct output_stream_buffer *output_buffers,
+		int num_output_buffers)
+{
+	struct ModuleAdapter *mod_adp = (struct ModuleAdapter *) mod->priv.module_adapter;
+
+	return process_internal(mod_adp, input_buffers, num_input_buffers,
+				output_buffers, num_output_buffers);
+}
+
+
+static int adp_wrapper_set_configuration(struct processing_module *mod,
+		 uint32_t config_id,
+		 enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+		 const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+		 size_t response_size)
+{
+	struct ModuleAdapter *mod_adp = (struct ModuleAdapter *) mod->priv.module_adapter;
+
+	return set_configuration_internal(mod_adp, config_id, pos, data_offset_size, fragment,
+			fragment_size, response, response_size);
+}
+
+/* Processing Module Adapter API*/
+static struct module_interface adp_interface = {
+	.init  = intel_modules_init,
+	.prepare = intel_modules_prepare,
+	.process = intel_modules_process,
+	.set_processing_mode = adp_wrapper_set_processing_mode,
+	.get_processing_mode = adp_wrapper_get_processing_mode,
+	.set_configuration = adp_wrapper_set_configuration,
+	.reset = adp_wrapper_reset,
+	.free = intel_modules_free,
+};
+
+/**
+ * \brief Init comp_driver data.
+ * \param[in] drv - component driver pointer.
+ *
+ * \note: For dynamically loaded module the comp_driver data are not yet initialized.
+ *        They need to be set in runtime, since base FW get uuid and module information
+ *        during load time.
+ */
+static void intel_modules_init_drv(struct comp_driver *drv)
+{
+	drv->ops.create2 = *intel_modules_shim_new;
+	drv->ops.prepare = module_adapter_prepare;
+	drv->ops.params = module_adapter_params;
+	drv->ops.copy = module_adapter_copy;
+	drv->ops.cmd = module_adapter_cmd;
+	drv->ops.trigger = module_adapter_trigger;
+	drv->ops.reset = module_adapter_reset;
+	drv->ops.free = module_adapter_free;
+}
+
+/**
+ * \brief Create a module adapter component.
+ * \param[in] drv - component driver pointer.
+ * \param[in] config - component ipc descriptor pointer.
+ * \param[in] spec - pointer to module configuration data
+ * \param[in] spec_size - size of module configuration data in dwords
+ *
+ * \return: a pointer to newly created module adapter component on success. NULL on error.
+ *
+ * \note: For dynamically loaded module the spec size is not known by base FW, since this is
+ *        loaded module specific information. Therefore configuration size is required here.
+ *        New module details are discovered during its loading, therefore comp_driver initialisation
+ *        happens at this point.
+ */
+struct comp_dev *intel_modules_shim_new(const struct comp_driver *drv,
+			struct comp_ipc_config *config,
+			void *spec, uint32_t spec_size)
+{
+
+	intel_modules_init_drv((struct comp_driver *)drv);
```

Comment:
instead of this can we add the ops in the DECLARE_DYNAMIC_MODULE_ADAPTER() macro?

Reply:
Yes, this is good idea. Will fix it.

Reply:
fixed